plugin/documents.py

```diff
@@ -959,6 +974,32 @@ def _update_stored_selection_async(self) -> Tuple[Optional[sublime.Region], bool
         self._stored_selection = selection
         return changed_first_region, True
 
+    def _format_on_paste_async(self) -> None:
+        clipboard_text = sublime.get_clipboard()
+        sel = self.view.sel()
+        split_clipboard_text = clipboard_text.split('\n')
+        multi_cursor_paste = len(split_clipboard_text) == len(sel) and len(sel) > 1
+        original_selection = list(sel)
+        regions_to_format = []  # type: List[sublime.Region]
+        pasted_text = clipboard_text
+        # add regions to selection, in order for lsp_format_document_range to format those regions
+        for index, region in enumerate(sel):
+            if multi_cursor_paste:
+                pasted_text = split_clipboard_text[index]
+            pasted_region = self.view.find(pasted_text, region.end(), sublime.REVERSE | sublime.LITERAL)
+            if pasted_region:
+                # Including whitespace may help servers format a range better
+                # More info at https://github.com/sublimelsp/LSP/pull/2311#issuecomment-1688593038
+                a = self.view.find_by_class(pasted_region.a, False,
+                                            sublime.CLASS_WORD_END | sublime.CLASS_PUNCTUATION_END)
+                formatting_region = sublime.Region(a, pasted_region.b)
+                regions_to_format.append(formatting_region)
+        sel.add_all(regions_to_format)
+        self.purge_changes_async()
+        self.view.run_command('lsp_format_document_range')
+        sel.clear()
+        sel.add_all(original_selection)
```

Comment:
Hmm, I was pretty sure that ST always runs commands on the main thread... If that would be the case then you would restore the original selection before the request would actually be sent. Can you verify whether this works as expected?

Reply:
I put a print statements:
```py
        self.purge_changes_async()
        self.view.run_command('lsp_format_document_range')
        print('restore')
        sel.clear()
        sel.add_all(original_selection)
# ...

class LspFormatDocumentRangeCommand(LspTextCommand):
    def run(self, edit: sublime.Edit, event: Optional[dict] = None) -> None:
        print('command invoked')
        if has_single_nonempty_selection(self.vi
```

first is the command invoked then the selection is restored:
```
command invoked
restore
```

It looks ok to me.

Reply:
Didn't realize before but ST appears to be blocking the calling (async) thread while the command runs on the main thread. So this acts like a sync code more or less.

I always thought that it schedules a task on the main thread.